packages/console/src/pages/SignInExperience/components/ManageLanguageModal/index.tsx

```diff
@@ -39,20 +41,70 @@ const ManageLanguageModal = ({ isOpen, onClose }: ManageLanguageModalProps) => {
     [customPhraseResponses]
   );
 
-  const defaultLanguageTag = allLanguageTags[0] ?? 'en';
+  const [displayedLanguages, setDisplayedLanguages] = useState<LanguageTag[]>(existedTags);
+
+  const defaultLanguageTag = existedTags[0] ?? 'en';
 
   const [selectedLanguageTag, setSelectedLanguageTag] = useState<LanguageTag>(defaultLanguageTag);
 
+  useEffect(() => {
+    setDisplayedLanguages(existedTags);
+  }, [existedTags]);
+
   const [isLanguageEditorDirty, setIsLanguageEditorDirty] = useState(false);
 
-  const [isUnsavedAlertOpen, setIsUnsavedAlertOpen] = useState(false);
+  const [confirmationState, setConfirmationState] = useState<ConfirmationState>('none');
   const [preselectedLanguageTag, setPreselectedLanguageTag] = useState<LanguageTag>();
+  const [preAddedLanguageTag, setPreAddedLanguageTag] = useState<LanguageTag>();
+  const [isAddingCustomLanguage, setIsAddingCustomLanguage] = useState(false);
+
+  const addLanguage = (languageTag: LanguageTag) => {
+    setDisplayedLanguages([...new Set([languageTag, ...existedTags])].slice().sort());
+    setSelectedLanguageTag(languageTag);
+    setIsAddingCustomLanguage(true);
+  };
+
+  const leaveAddingCustomLanguageState = () => {
+    setDisplayedLanguages(existedTags);
+  };
 
   useEffect(() => {
     if (!isOpen) {
       setSelectedLanguageTag(defaultLanguageTag);
     }
-  }, [allLanguageTags, setSelectedLanguageTag, isOpen, defaultLanguageTag]);
+  }, [setSelectedLanguageTag, isOpen, defaultLanguageTag]);
+
+  const handleConfirmClose = () => {
+    if (isAddingCustomLanguage) {
+      leaveAddingCustomLanguageState();
+      setIsAddingCustomLanguage(false);
+    }
+
+    onClose();
+  };
+
+  const handleConfirmSwitchLanguage = () => {
+    if (preselectedLanguageTag) {
+      setSelectedLanguageTag(preselectedLanguageTag);
+      setPreselectedLanguageTag(undefined);
+    }
+
+    if (isAddingCustomLanguage) {
+      leaveAddingCustomLanguageState();
+      setIsAddingCustomLanguage(false);
+    }
+  };
+
+  const handleConfirmAddLanguage = () => {
+    if (isAddingCustomLanguage) {
+      leaveAddingCustomLanguageState();
+      setIsAddingCustomLanguage(false);
+    }
+
+    if (preAddedLanguageTag) {
+      addLanguage(preAddedLanguageTag);
+    }
+  };
```

Comment:
Looks like some duplicate logics. Could we use a single onConfirm callback based on different request type?

Reply:
updated.